vortex-datafusion/src/eval.rs

```diff
@@ -7,35 +7,84 @@ use vortex::{
+        debug_assert!(can_be_pushed_down(expr));
+        let original_len = array.len();
         match expr {
             Expr::BinaryExpr(expr) => {
-                let lhs = ExpressionEvaluator::eval(array.clone(), expr.left.as_ref())?;
-                let rhs = ExpressionEvaluator::eval(array, expr.right.as_ref())?;
-
                 // TODO(adamg): turn and/or into more general compute functions
                 match expr.op {
                     DFOperator::And => {
+                        let lhs = ExpressionEvaluator::eval(array.clone(), expr.left.as_ref())?;
                         let lhs = lhs.into_bool()?;
+
+                        if lhs.true_count() == 0 {
+                            return Ok(ConstantArray::new(false, original_len).into_array());
+                        }
+
+                        let rhs = ExpressionEvaluator::eval(array, expr.right.as_ref())?;
                         let rhs = rhs.into_bool()?;
+
+                        if rhs.true_count() == 0 {
+                            return Ok(ConstantArray::new(false, original_len).into_array());
+                        }
+
                         let buffer = &lhs.boolean_buffer() & &rhs.boolean_buffer();
                         Ok(BoolArray::from(buffer).into_array())
                     }
                     DFOperator::Or => {
+                        let lhs = ExpressionEvaluator::eval(array.clone(), expr.left.as_ref())?;
                         let lhs = lhs.into_bool()?;
+
+                        if lhs.true_count() == original_len {
+                            return Ok(ConstantArray::new(true, original_len).into_array());
+                        }
+
+                        let rhs = ExpressionEvaluator::eval(array, expr.right.as_ref())?;
                         let rhs = rhs.into_bool()?;
+
+                        if lhs.true_count() == original_len {
+                            return Ok(ConstantArray::new(true, original_len).into_array());
+                        }
+
                         let buffer = &lhs.boolean_buffer() | &rhs.boolean_buffer();
                         Ok(BoolArray::from(buffer).into_array())
                     }
-                    DFOperator::Eq => compare(&lhs, &rhs, Operator::Eq),
-                    DFOperator::Gt => compare(&lhs, &rhs, Operator::Gt),
-                    DFOperator::GtEq => compare(&lhs, &rhs, Operator::Gte),
-                    DFOperator::Lt => compare(&lhs, &rhs, Operator::Lt),
-                    DFOperator::LtEq => compare(&lhs, &rhs, Operator::Lte),
-                    DFOperator::NotEq => compare(&lhs, &rhs, Operator::NotEq),
+                    DFOperator::Eq => {
+                        let lhs = ExpressionEvaluator::eval(array.clone(), expr.left.as_ref())?;
+                        let rhs = ExpressionEvaluator::eval(array, expr.right.as_ref())?;
+                        compare(&lhs, &rhs, Operator::Eq)
+                    }
+                    DFOperator::Gt => {
```

Comment:
You're gonna want that `From<DFOperator>` impl!

Reply:
I think we should rethink the whole `Operator` thing here, if its more of an `Operation` we can add boolean operations and get rid of the negation logic.